cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -2513,6 +2877,159 @@ void AddSplit(FunctionRegistry* registry) {
+      auto num_repeats = input2.GetView(i);
+      if (num_repeats < 0) {
+        *st = InvalidRepeatCount();
+        return num_repeats;
+      }
+      total_num_repeats += num_repeats;
+    }
+    return input1_ncodeunits * total_num_repeats;
+  }
+
+  int64_t MaxCodeunits(const ArrayType1& input1, const int64_t num_repeats,
+                       Status*) override {
+    return input1.total_values_length() * num_repeats;
+  }
+
+  int64_t MaxCodeunits(const ArrayType1& input1, const ArrayType2& input2,
+                       Status* st) override {
+    int64_t total_codeunits = 0;
+    for (int64_t i = 0; i < input2.length(); ++i) {
+      auto num_repeats = input2.GetView(i);
+      if (num_repeats < 0) {
+        *st = InvalidRepeatCount();
+        return num_repeats;
+      }
+      total_codeunits += input1.GetView(i).length() * num_repeats;
+    }
+    return total_codeunits;
+  }
+
+  std::function<int64_t(const uint8_t*, const int64_t, const int64_t, uint8_t*, Status*)>
+      Transform;
+
+  static int64_t TransformSimple(const uint8_t* input,
+                                 const int64_t input_string_ncodeunits,
+                                 const int64_t num_repeats, uint8_t* output, Status*) {
+    uint8_t* output_start = output;
+    for (int64_t i = 0; i < num_repeats; ++i) {
+      std::memcpy(output, input, input_string_ncodeunits);
+      output += input_string_ncodeunits;
+    }
+    return output - output_start;
+  }
+
+  static int64_t TransformDoubling(const uint8_t* input,
+                                   const int64_t input_string_ncodeunits,
+                                   const int64_t num_repeats, uint8_t* output, Status*) {
+    uint8_t* output_start = output;
+    // Repeated doubling of string
+    std::memcpy(output, input, input_string_ncodeunits);
+    output += input_string_ncodeunits;
+    int64_t i = 1;
+    for (int64_t ilen = input_string_ncodeunits; i <= (num_repeats / 2);
+         i *= 2, ilen *= 2) {
+      std::memcpy(output, output_start, ilen);
+      output += ilen;
+    }
+
+    // Epilogue remainder
+    int64_t rem = (num_repeats ^ i) * input_string_ncodeunits;
+    std::memcpy(output, output_start, rem);
+    output += rem;
+    return output - output_start;
+  }
+
+  static int64_t TransformWrapper(const uint8_t* input,
+                                  const int64_t input_string_ncodeunits,
+                                  const int64_t num_repeats, uint8_t* output,
+                                  Status* st) {
+    auto transform = (num_repeats < 4) ? TransformSimple : TransformDoubling;
+    return transform(input, input_string_ncodeunits, num_repeats, output, st);
+  }
+
+  Status PreExec(KernelContext*, const ExecBatch& batch, Datum*) override {
+    // For cases with a scalar repeat count, select the best implementation once
+    // before execution. Otherwise, use TransformWrapper to select implementation
+    // when processing each value.
```

Comment:
I'm curious: is it really better to go through the `std::function` indirection than to compute `num_repeats < 4` at each iteration?

Reply:
I did not measured this so will run benchmarks to compare.

Reply:
Using `std::function` indirection resulted in 2x slower, so good call/intuition on this one.
```
StringRepeat_mean    622822087 ns    622817699 ns           10 bytes_per_second=25.4509M/s items_per_second=1.68498M/s
StringRepeat_median  623393064 ns    623390528 ns           10 bytes_per_second=25.4067M/s items_per_second=1.68205M/s
StringRepeat_stddev   18771545 ns     18770743 ns           10 bytes_per_second=787.511k/s items_per_second=50.9153k/s
```
Checking `num_repeats < 4` at each iteration
```
StringRepeat_mean    313125674 ns    313123902 ns           10 bytes_per_second=50.601M/s items_per_second=3.35004M/s
StringRepeat_median  312795031 ns    312794088 ns           10 bytes_per_second=50.6405M/s items_per_second=3.35266M/s
StringRepeat_stddev    6484104 ns      6484645 ns           10 bytes_per_second=1068k/s items_per_second=69.0502k/s
```